### ISSUE TITLE
Implement swim lanes

### ISSUE THREAD

**Issue by samcaulfield** [samcaulfield/pml-studio#29]
This feature is on the product backlog. It is worth 30 points.

Description:
> Actions should be allocated to "swim lanes" where each lane is a different agent.

**Comment by samcaulfield** [comment]
Have you managed to make much progress on this Eamon?

**Comment by EamonMc** [comment]
Hi Sam. No I haven't. I simply must go all-out today and tomorrow for my FYP demo on Tuesday. After that I will work every waking hour on pushing out features for group.

**Comment by sharonluk** [comment]
we're making progress today, outputting diagrams using PlantUML. Just need to convert it to traverse.

**Comment by EamonMc** [comment]
Putting in ridiculous hours at this - back on it after some sleep. His new model is particularly tricky. Will get there though.

**Comment by samcaulfield** [comment]
OK. Just to note that it's risky to make any major changes to the code this weekend. As has happened to other teams, major changes introduced at the last minute can break the system and actually cost marks. Do you think you could have this working today?

**Comment by EamonMc** [comment]
Yes - going at this for next 12hrs. I'll be on here and fb chat if you have time to look at it / test it when committed.

**Comment by EamonMc** [comment]
Ok the bulk of the translating and code is now committed. Thanks for the help @sharonluk 

This is just the initial commit - the logic of the graph will be done after.

It basically builds a string in plantUML format.

A new version of traverse named 'simpletraverse' is in the thirdparty/simpleTraverse folder along with the simpleTraverse.c that it was compiled from. This outputs a nicer language to parse than DOT output. 
The parsing is done in script.js (rather than traverse.c)
In this function ( swimlanesBuilder()) there is now some dictionaries and arrays of the node elements
dictNodeAgents ->        nodeName : [ agent1, agent2,...]
dictNodeType   ->         nodeName : nodeType  
and all graph connections as in DOT as a 2D array of connections:
twoDArrayConnections[][] = [ [node1, node2] , [node3, node8]...]

We can possibly use this instead of visjs for the main diagram also. I'll work on the graph logic now and how to represent branches / iterations/ selections etc in the plantUML string builder

**Comment by samcaulfield** [comment]
This looks promising.

**Comment by EamonMc** [comment]
We are still working on getting the branch/select logic working. We have a recursive loop that traverses the graph building the new one but haven't quite got the select/branch logic correct. Worried because has to be "insanely great" by 12. Keeping on this.

**Comment by EamonMc** [comment]
This was a valiant effort by me and Sharon. I would like to think we'll get the points for this as it was quite ambitious. Some of the visual strangeness in swimlane flow control arrows stem from the plantUML API not handling swimlanes and splits well and can not be improved on. (removing the swimlanes fixes it) 

There is also now a button for 'visualise plantUML' as another alternative to the visjs diagram which has its own failings.

It stops on netbeans because I believe traverse sees the first SELECTION as a JOIN for some reason and this messes  up the logic. I'll try patch this but is a lot to check every JOIN to see if it is really a SELCTION.
This can be seen in the graphviz diagram also as the split comes from a join.
I'll see if I can make some minor improvements before the deadline.

**Comment by EamonMc** [comment]
OMFG I thought it was sunday today :+1:  Troy just explained to me that it's in-fact Saturday. This is insane. Have another 24hrs to sort this.

**Comment by samcaulfield** [comment]
This is shaping up to be a really great feature. When you're polishing it off at the end you can try adding a [scrollbar](http://www.w3schools.com/cssref/pr_pos_overflow.asp) to the graphical editor to handle overflow.
